crates/flux-infer/src/fixpoint_encoding.rs

```diff
@@ -1064,10 +1072,27 @@ impl<'genv, 'tcx> ExprEncodingCtxt<'genv, 'tcx> {
                 };
                 self.expr_to_fixpoint(&expr, scx)
             }
-            InternalFuncKind::CharToInt | InternalFuncKind::IntToChar => {
-                // We can erase the "call" because fixpoint uses `int` to represent `char`
-                // so the conversions are unnecessary.
-                self.expr_to_fixpoint(&args[0], scx)
+            InternalFuncKind::Cast => {
+                match sort_args {
+                    [rty::SortArg::Sort(from), rty::SortArg::Sort(to)] if is_id_cast(from, to) => {
+                        self.expr_to_fixpoint(&args[0], scx)
+                    }
+                    [rty::SortArg::Sort(rty::Sort::Bool), rty::SortArg::Sort(rty::Sort::Int)] => {
+                        let expr = rty::Expr::ite(
+                            args[0].clone(),
+                            rty::Expr::constant(rty::Constant::from(1)),
+                            rty::Expr::constant(rty::Constant::from(0)),
+                        );
+                        self.expr_to_fixpoint(&expr, scx)
+                    }
+                    // Treat all other to_int conversions as uninterpreted
```

Comment:
I'm not a fan of casts being silently "ignored" as uninterpreted functions. That's a recipe for "why doesn't my code verify?"

But on the plus side, if everything is uniformly castable, then you can have, for example, a function generic on a sort that is castable to `int` without having explicit bounds.

Maybe we can have the cake and eat it too, if we reject statically casts that don't make sense when both sides are concrete.

Reply:
Yes, they are genuine use cases for the uif. The problem is when it's unintended.

This could be implemented as a warn-by-default lint that can be silenced. But we don't have any other examples of that so I don't know how hard that would be.

Reply:
How about we have error-by-default but you instead can have a flag to `allow_uninterpreted_cast` ?

Reply:
Yes, that would have the intended behavior. We could even make it per item if we add it as an option to `InferOpts`.

Reply:
Also, there are two ways in which one could fail. 

1. During sort checking
2. During fixpoint encoding

I'm thinking about 1.